sdk/core/azure-core/inc/azure/core/internal/hkeyholder.hpp

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Internal utility functions for strings.
+ *
+ */
+#pragma once
+
+#if defined(AZ_PLATFORM_WINDOWS)
+#if !defined(WIN32_LEAN_AND_MEAN)
+#define WIN32_LEAN_AND_MEAN
+#endif
+#if !defined(NOMINMAX)
+#define NOMINMAX
+#endif
+
+#include <windows.h>
+
+namespace Azure { namespace Core { namespace Internal {
+
+  class HKeyHolder {
+  private:
+    HKEY m_value;
+
+  public:
+    HKeyHolder() : m_value(NULL) {}
+
+    ~HKeyHolder()
+    {
+      if (m_value != NULL)
+      {
+        ::RegCloseKey(m_value);
+      }
+    }
+
+    void operator=(HKEY p)
+    {
+      if (p != NULL)
+      {
+        m_value = p;
```

Comment:
Should we close the previous `m_value` before overriding it during assignment?

Otherwise, what happens with this:
```C++
{
  Azure::Core::Internal::HKeyHolder key1 = <some registry key 1>;
  Azure::Core::Internal::HKeyHolder key2 = <some registry key 2>;

  key2 = key1; // Who will now close <some registry key 2>
}
// When we leave out of scope, the destructor for key2 gets called, which
// closes <some registry key 1>
```

Reply:
We could enhance this to not allow assignment or we could make the assignment close the handle.  I can look at that in a future change